clippy_lints/src/unnecessary_semicolon.rs

```diff
@@ -48,6 +93,23 @@ impl LateLintPass<'_> for UnnecessarySemicolon {
             )
             && cx.typeck_results().expr_ty(expr) == cx.tcx.types.unit
         {
+            if self.is_last_in_block(stmt)
+                && for_each_unconsumed_temporary(cx, expr, |temporary_ty| {
+                    if temporary_ty.has_significant_drop(cx.tcx, cx.typing_env())
+                        && temporary_ty
+                            .walk()
+                            .any(|arg| matches!(arg.unpack(), GenericArgKind::Lifetime(re) if !re.is_static()))
+                    {
+                        ControlFlow::Break(())
+                    } else {
+                        ControlFlow::Continue(())
+                    }
+                })
+                .is_break()
```

Comment:
Suggestion: now that there are two lints that need to do this it probably makes sense to move this to clippy_utils/ty, since I can see how we might want to adjust the logic in here a little bit in the future

Reply:
Done. The name is a bit of a mouthful, but I reflects what the function is doing (and it is quite a specialized one).